lightning/src/blinded_path/payment.rs

```diff
@@ -0,0 +1,288 @@
+//! Data structures and methods for constructing [`BlindedPath`]s to send a payment over.
+//!
+//! [`BlindedPath`]: crate::blinded_path::BlindedPath
+
+use bitcoin::secp256k1::{self, PublicKey, Secp256k1, SecretKey};
+use core::convert::TryFrom;
+use crate::blinded_path::BlindedHop;
+use crate::blinded_path::utils;
+use crate::io;
+use crate::ln::PaymentSecret;
+use crate::ln::features::BlindedHopFeatures;
+use crate::ln::msgs::DecodeError;
+use crate::offers::invoice::BlindedPayInfo;
+use crate::prelude::*;
+use crate::util::ser::{Readable, Writeable, Writer};
+
+/// Data to construct a [`BlindedHop`] for sending a payment over.
+///
+/// [`BlindedHop`]: crate::blinded_path::BlindedHop
+pub enum BlindedPaymentTlvs {
+	/// This blinded payment data is to be forwarded.
+	Forward {
+		/// The short channel id this payment is being forwarded over.
+		short_channel_id: u64,
+		/// Payment parameters for relaying over this channel.
+		payment_relay: PaymentRelay,
+		/// Payment constraints when relaying over this channel.
+		payment_constraints: PaymentConstraints,
+		/// Supported and required features when relaying a payment onion containing this object's
+		/// corresponding [`BlindedHop`].
+		///
+		/// [`BlindedHop`]: crate::blinded_path::BlindedHop
+		features: BlindedHopFeatures,
+	},
+	/// This blinded payment data is to be received.
+	Receive {
+		/// Used to authenticate the sender of a payment to the receiver and tie MPP HTLCs together.
+		payment_secret: PaymentSecret,
```

Comment:
Note for reviewers: it's a bit LDK-specific and not generic that we take a `PaymentSecret` here. But as discussed previously with @TheBlueMatt, the spec should not be prescribing anything for the final payload at all because it's constructed for the receiver, by the receiver, so I don't see an issue with doing this instead of calling this field `path_id` and taking a `Vec<u8>` (which is an unnecessary allocation as well). But I get that this might be a bit confusing coming from the spec so thoughts welcome here.

Reply:
Where is this encoded in the `BlindedPath`?

Reply:
Ah, I see. It's encoded in the `BlindedPath` when `utils::encrypt_payload` is called.